bpf/lib/encrypt.h

```diff
@@ -284,8 +288,109 @@ encrypt_overlay_and_redirect(struct __ctx_buff *ctx)
 	return ret;
 }
 #endif /* ENABLE_ENCRYPTED_OVERLAY */
+#elif defined(ENABLE_WIREGUARD)
+
+/* Returns CTX_ACT_REDIRECT in case of a cilium-managed WireGuard packet.
+ * Differently than do_encrypt() for IPSec, WireGuard does not use the ctx->mark
+ * to indicate whether a packet needs decryption. Therefore, we can just return
+ * CTX_ACT_REDIRECT to the caller, or any other value in case of non-WireGuard packet.
+ * The conditions sequentially checked in order to establish that a WireGuard packet is
+ * valid and should be traced with TRACE_REASON_ENCRYPT are as follows:
+ *
+ * - UDP packet
+ * - L4 dport == WG_PORT
+ * - L4 sport == dport
+ * - Valid node ID
+ */
+static __always_inline int
+do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+{
+	void *data, *data_end;
+	__u8 protocol = 0;
+	__u16 node_id = 0;
+	int l4_off = 0;
+	struct {
+		__be16 sport;
+		__be16 dport;
+	} l4;
+#ifdef ENABLE_IPV6
+	struct ipv6hdr *ip6;
+#endif
+#ifdef ENABLE_IPV4
+	struct iphdr *ip4;
+#endif
 
-#else
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))
+			return CTX_ACT_OK;
+		protocol = ip6->nexthdr;
+		l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &protocol);
+		break;
+#endif
+#ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data_pull(ctx, &data, &data_end, &ip4))
+			return CTX_ACT_OK;
+		protocol = ip4->protocol;
+		l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
+		break;
+#endif
+	default:
+		return CTX_ACT_OK;
+	}
+
+	/* Ignore non-UDP packets. */
+	if (protocol != IPPROTO_UDP)
+		return CTX_ACT_OK;
+
+	/* Ignore if unable to retrieve L4 ports. */
+	if (l4_load_ports(ctx, l4_off + UDP_SPORT_OFF, &l4.sport) < 0)
+		return send_drop_notify_error(ctx, UNKNOWN_ID, DROP_INVALID,
+						  CTX_ACT_DROP,
+						  METRIC_INGRESS);
+
+	/* Ignore if packet is not for cilium@WireGuard*/
+	if (l4.dport != bpf_htons(WG_PORT))
+		return CTX_ACT_OK;
+
+	/* Ignore if packet not probably originated by cilium@WireGuard*/
+	if (l4.sport != l4.dport)
+		return send_drop_notify_error(ctx, UNKNOWN_ID, DROP_INVALID,
+						  CTX_ACT_DROP,
+						  METRIC_INGRESS);
+
+	/* Late retrieval of node ID. Here we are sure the packet
+	 * comes and is destined to cilium@WireGuard ports, therefore
+	 * it must be originated by a valid cilium-managed node.
+	 */
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		node_id = lookup_ip6_node_id((union v6addr *)&ip6->saddr);
+		break;
+#endif
+#ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		node_id = lookup_ip4_node_id(ip4->saddr);
+		break;
+#endif
+	default:
+		return CTX_ACT_OK;
+	}
+
+	/* Ignore invalid node ID. */
+	if (!node_id)
```

Comment:
Ipsec needs node id to decrypt but not for wireguard

Reply:
I wanted to make sure that the sender was a valid node.
Would you then just check for dport==WG_PORT? Or also make sure that sport==dport?

Reply:
Can we move the whole "check wireguard and notify" to https://github.com/cilium/cilium/blob/v1.17.0-pre.1/bpf/bpf_host.c#L593 in the handle_ipv4/handle_ipv6, where we have the ipcache_srcid and ip4/ip6 parsed. Over there we can call `identity_is_cluster(ipcache_srcid)` to make sure the packet is from inside the cilium.

Reply:
Let me investigate this solution. Though, at this point `bpf_host` would have already traced `FROM_NETWORK` without the TRACE_REASON_ENCRYPTED, so we're creating a new tracing here?

Reply:
Good point. Therefore `do_netdev()` indeed is better choice. 
Let's just use the src port and dst port to distinguish wireguard.